pkg/sdk/functions_gen_test.go

```diff
@@ -0,0 +1,549 @@
+		opts.ReturnResultsBehavior = ReturnResultsBehaviorPointer(ReturnResultsBehaviorImmutable)
+		opts.RuntimeVersion = String("2.0")
+		opts.Comment = String("comment")
+		opts.Imports = []FunctionImports{
+			{
+				Import: "@udf_libs/echohandler.jar",
+			},
+		}
+		opts.Handler = "Echo.echoVarchar"
+		opts.FunctionDefinition = String("return x")
+		assertOptsValidAndSQLEquals(t, opts, `CREATE OR REPLACE TEMPORARY SECURE FUNCTION %s (x VARCHAR DEFAULT 'test') COPY GRANTS RETURNS VARCHAR NOT NULL LANGUAGE SCALA CALLED ON NULL INPUT IMMUTABLE RUNTIME_VERSION = '2.0' COMMENT = 'comment' IMPORTS = ('@udf_libs/echohandler.jar') HANDLER = 'Echo.echoVarchar' AS 'return x'`, id.FullyQualifiedName())
+	})
+}
+
+func TestFunctions_CreateForSQL(t *testing.T) {
+	id := RandomSchemaObjectIdentifier()
+
+	defaultOpts := func() *CreateForSQLFunctionOptions {
+		return &CreateForSQLFunctionOptions{
+			name: id,
+		}
+	}
+
+	t.Run("validation: nil options", func(t *testing.T) {
+		opts := (*CreateForSQLFunctionOptions)(nil)
+		assertOptsInvalidJoinedErrors(t, opts, ErrNilOptions)
+	})
+
+	t.Run("validation: incorrect identifier", func(t *testing.T) {
+		opts := defaultOpts()
+		opts.name = NewSchemaObjectIdentifier("", "", "")
+		assertOptsInvalidJoinedErrors(t, opts, ErrInvalidObjectIdentifier)
+	})
+
+	t.Run("all options", func(t *testing.T) {
+		opts := defaultOpts()
+		opts.OrReplace = Bool(true)
+		opts.Temporary = Bool(true)
+		opts.Secure = Bool(true)
+		opts.Arguments = []FunctionArgument{
+			{
+				ArgName:      "message",
+				ArgDataType:  "VARCHAR",
+				DefaultValue: String("'test'"),
+			},
+		}
+		opts.CopyGrants = Bool(true)
+		opts.Returns = FunctionReturns{
+			ResultDataType: &FunctionReturnsResultDataType{
+				ResultDataType: DataTypeFloat,
+			},
+		}
+		opts.ReturnNullValues = ReturnNullValuesPointer(ReturnNullValuesNotNull)
+		opts.ReturnResultsBehavior = ReturnResultsBehaviorPointer(ReturnResultsBehaviorImmutable)
+		opts.Memoizable = Bool(true)
+		opts.Comment = String("comment")
+		opts.FunctionDefinition = String("3.141592654::FLOAT")
+		assertOptsValidAndSQLEquals(t, opts, `CREATE OR REPLACE TEMPORARY SECURE FUNCTION %s (message VARCHAR DEFAULT 'test') COPY GRANTS RETURNS FLOAT NOT NULL IMMUTABLE MEMOIZABLE COMMENT = 'comment' AS '3.141592654::FLOAT'`, id.FullyQualifiedName())
+	})
+}
+
+func TestFunctions_Drop(t *testing.T) {
+	id := RandomSchemaObjectIdentifier()
+
+	defaultOpts := func() *DropFunctionOptions {
+		return &DropFunctionOptions{
+			name: id,
+		}
+	}
+
+	t.Run("validation: nil options", func(t *testing.T) {
+		opts := (*DropFunctionOptions)(nil)
+		assertOptsInvalidJoinedErrors(t, opts, ErrNilOptions)
+	})
+
+	t.Run("validation: incorrect identifier", func(t *testing.T) {
+		opts := defaultOpts()
+		opts.name = NewSchemaObjectIdentifier("", "", "")
+		assertOptsInvalidJoinedErrors(t, opts, ErrInvalidObjectIdentifier)
+	})
+
+	t.Run("all options", func(t *testing.T) {
```

Comment:
done

Reply:
You have added a test for empty list, but no params are generated:
```
	t.Run("all options no arguments", func(t *testing.T) {
		opts := defaultOpts()
		assertOptsValidAndSQLEquals(t, opts, `DESCRIBE FUNCTION %s`, id.FullyQualifiedName())
	})
```
The documentation states:
```
DESC[RIBE] FUNCTION <name> ( [ <arg_data_type> ] [ , ... ] )
```

IMO this means, that parentheses should be generated even if there are no arguments (and your test proves otherwise).

Can you add an integration test confirming how it really works? (because it's either problem with the docs or with the generated SQL)

Reply:
test fixed, and integration test added

Reply:
I still can't see:
- test for generating a proper DESCRIBE for no argument function (it was there earlier, but I guess you deleted it on purpose or accidentally)
- test for generating a proper DROP for no argument function
- integration test for describe no argument function (the one you have added here: https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/2205/files#diff-80d45f48c5a2ca18b97aa742a5c9e9c90bfc904d6646a3547a3dc2133bc0e3b8R175 runs only SHOW, no describe)

Please add them, so we can finally close it